src/Storages/StorageMaterializedView.cpp

```diff
@@ -139,15 +141,27 @@ void StorageMaterializedView::read(
 {
     auto storage = getTargetTable();
     auto lock = storage->lockForShare(context.getCurrentQueryId(), context.getSettingsRef().lock_acquire_timeout);
-    auto metadata_snapshot = storage->getInMemoryMetadataPtr();
+    auto target_metadata_snapshot = storage->getInMemoryMetadataPtr();
 
     if (query_info.order_optimizer)
-        query_info.input_order_info = query_info.order_optimizer->getInputOrder(metadata_snapshot, context);
+        query_info.input_order_info = query_info.order_optimizer->getInputOrder(target_metadata_snapshot, context);
 
-    storage->read(query_plan, column_names, metadata_snapshot, query_info, context, processed_stage, max_block_size, num_streams);
+    storage->read(query_plan, column_names, target_metadata_snapshot, query_info, context, processed_stage, max_block_size, num_streams);
 
     if (query_plan.isInitialized())
     {
+        auto mv_header = getHeaderForProcessingStage(*this, column_names, metadata_snapshot, query_info, context, processed_stage);
```

Comment:
It seems that we can get `query_plan.getHeader()` instead of `mv_header` cause `query_plan.addStep` checks that headers are equal

Reply:
Do you mean `query_plan.getCurrentDataStream().header`? If I understand correctly, `query_plan.getCurrentDataStream().header` should be equal to `target_header` (not `mv_header`) because query plan is initialized by target storage.

Reply:
Ah, yes, indeed :)